### PR TITLE
bridge: add isDefaultGateway field

### LINKED ISSUE #27: bridge: does not set the gateway on the route
First raised in https://github.com/coreos/rkt/issues/862

bridge does correctly sets the IP on on the bridge (if `isGateway` is true) but not on the route.

### LINKED ISSUE #184: flannel plugin should allow setting default routes
Currently the flannel plugin, when generating config for the `delegate` IPAM seems to be setting only routes to the flannel network. However, it should be setting up a default route as well. Otherwise containers launched using the flannel plugin cannot communicate with anything other than containers within the flannel network.

### PR DESCRIPTION
**Pull request by steveeJ**
When isDefaultGateway is true it automatically sets isGateway to true.
The default route will be added via the (bridge's) gateway IP.
If a default gateway has been configured via IPAM in the same configuration file, the plugin will error out.

Fixes #27.
Fixes #184.

### CODE REVIEW

File: plugins/main/bridge/bridge.go [hunk 1]
```diff
@@ -196,6 +201,28 @@ func cmdAdd(args *skel.CmdArgs) error {
...
+	// set the default gateway if requested
+	if n.IsDefaultGW {
+		_, defaultNet, err := net.ParseCIDR("0.0.0.0/0")
+		if err != nil {
+			return err
+		}
+		for i, route := range result.IP4.Routes {
+			if defaultNet.String() == route.Dst.String() {
+				if route.GW != nil && !route.GW.Equal(result.IP4.Gateway) {
+					return errors.New(fmt.Sprintf(
```
Comment (eyakubovich) [inline plugins/main/bridge/bridge.go]: `fmt.Errorf`

**Comment by eyakubovich** [comment]
I think it's a reasonable solution but I need to sleep on this. There are some implications for integrating it with `rkt` and it setting up a default route.

**Comment by steveeJ** [comment]
I'd like to point out again that the default behaviour hasn't been changed, it does require a configuration change to make this happen. Only if `isDefaultGateway` is provided *instead of* or in addition to `isGateway` the default gateway will be set.

**Comment by eyakubovich** [comment]
@steveeJ I guess you're right. However there's also fact that this almost sugar on top of IPAM routes. You could do the same with `"routes": { "dst": "0.0.0.0/0", "gw": "$GW }` in the IPAM section. At the same this maybe inconvenient but in that case it should be handled in the IPAM section/plugins and not the bridge.

**Comment by steveeJ** [comment]
I see your point. I followed the already existing `isGateway`, which could also replaced via IPAM by adding a route that has the IPAM subnet as the destination.
One thing that has bothered me so far is that the gateway is automatically assumed to be the first subnet address using the `isGateway` attribute. In case the gateway is different, people would have to configure the routes manually via IPAM, and the `isGateway` would not fit in anymore.

So, I would like a solution of that moves the gateway attributes to IPAM and allows for better flexibility. 
Opinions?

**Comment by eyakubovich** [comment]
The first address is only assumed if IPAM did not provide a gateway address (https://github.com/appc/cni/blob/master/plugins/main/bridge/bridge.go#L195). The "gateway" field in the IPAM result JSON is optional, so there is a problem that if IPAM returns some routes but not the gateway, it can get messed up. Maybe we can issue a warning or something in this case.

**Comment by steveeJ** [comment]
@eyakubovich: my understanding on this is that we'll move the gateway attributes to IPAM and leave the behavior as is, but add a convenience option to set the default GW as already implemented in this PR.

Please confirm and I'll update the implementation accordingly.

File: plugins/main/bridge/bridge.go [hunk 2]
```diff
@@ -196,6 +201,7 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return err
 	}
 
+	// TODO: make this optional when IPv6 is supported
```
Comment (jellonek) [inline plugins/main/bridge/bridge.go]: After inclusion of IPv6 next line should be expanded with `&& result.IP6 == nil` so this comment is hmmm... not precise.

**Comment by jellonek** [comment]
LGTM

File: plugins/main/bridge/bridge.go (hunk 2 above)
Comment (steveeJ) [inline plugins/main/bridge/bridge.go]: It's a hint that there's going to be multiple options, logically `IPv4 || IPv6`

**Comment by squaremo** [comment]
LGTM. As discussed on IRC, it's worth waiting for #211 and rebasing, for the tests, before merging.

**Pull request merged**